### ISSUE TITLE
afl-clang-lto++: Conflicts between libstdc++ and libc++

### ISSUE THREAD

**Issue by WillyPillow** [AFLplusplus/AFLplusplus#865]
**IMPORTANT**
- [x] You have verified that the issue to be present in the current `dev` branch
- [x] Please supply the command line options and relevant environment variables, e.g. a copy-paste of the contents of `out/default/fuzzer_setup`

Thank you for making afl++ better!

**Describe the bug**

Building the following program with `afl-clang-lto++` results in a invalid pointer dereference, causing it to segfault.

``` cpp
#include <fstream>
#include <iostream>

int main() {
	std::fstream fs("/dev/null");
	fs.exceptions(std::ios_base::failbit);
	int x{};
	fs >> x;
	std::cerr << x << std::endl;
}
```

**To Reproduce**

1. Build the aforementioned program with `afl-clang-lto++ test.cpp`
2. Observe the segfault

**Expected behavior**

The program should abort with a `std::__ios_failure` exception.

**Screen output/Screenshots**

Stack trace:

```
#0  0x00007facf3c9cba2 in std::type_info::operator== (__arg=..., this=0x7facf3dc9d88 <typeinfo for std::__ios_failure>) at /build/gcc/src/gcc/libstdc++-v3/libsupc++/typeinfo:122
#1  __cxxabiv1::__si_class_type_info::__do_dyncast (this=0x7facf3dc9d88 <typeinfo for std::__ios_failure>, src2dst=140720525607144, access_path=823148896, dst_type=0x1, obj_ptr=0x1, src_type=0x0, src_ptr=0x7ffc0cf10578, result=...)
    at /build/gcc/src/gcc/libstdc++-v3/libsupc++/si_class_type_info.cc:52
#2  0x00007facf3e02f0a in ?? () from /usr/lib/libc++abi.so.1
#3  0x00007facf3de6344 in ?? () from /usr/lib/libc++abi.so.1
#4  0x00007facf3e01933 in ?? () from /usr/lib/libc++abi.so.1
#5  0x00007facf3e048f6 in ?? () from /usr/lib/libc++abi.so.1
#6  0x00007facf3e04871 in __cxa_throw () from /usr/lib/libc++abi.so.1
#7  0x00007facf3c92f12 in std::__throw_ios_failure (__s=__s@entry=0x7facf3d82807 "basic_ios::clear") at /build/gcc/src/gcc/libstdc++-v3/src/c++11/cxx11-ios_failure.cc:169
#8  0x00007facf3d07f03 in std::basic_ios<char, std::char_traits<char> >::clear (this=<optimized out>, __state=<optimized out>) at /build/gcc/src/gcc-build/x86_64-pc-linux-gnu/libstdc++-v3/include/bits/basic_ios.tcc:48
#9  0x0000558430ea4565 in std::basic_ios<char, std::char_traits<char> >::exceptions (this=0x7facf3dc9d88 <typeinfo for std::__ios_failure>, __except=std::_S_failbit) at /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/10.2.0/../../../../include/c++/10.2.0/bits/basic_ios.h:260
#10 main () at test.cpp:6
```

Address sanitizer:

```
AddressSanitizer:DEADLYSIGNAL
=================================================================
==347995==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000009 (pc 0x7fa83f349ba2 bp 0x6100000000c0 sp 0x7fffdd5b1b00 T0)
==347995==The signal is caused by a READ memory access.
==347995==Hint: address points to the zero page.
    #0 0x7fa83f349ba2 in std::type_info::operator==(std::type_info const&) const /build/gcc/src/gcc/libstdc++-v3/libsupc++/typeinfo:122:32
    #1 0x7fa83f349ba2 in __cxxabiv1::__si_class_type_info::__do_dyncast(long, __cxxabiv1::__class_type_info::__sub_kind, __cxxabiv1::__class_type_info const*, void const*, __cxxabiv1::__class_type_info const*, void const*, __cxxabiv1::__class_type_info::__dyncast_result&) const /build/gcc/src/gcc/libstdc++-v3/libsupc++/si_class_type_info.cc:52:13
    #2 0x7fa83f4aff09  (/usr/lib/libc++abi.so.1+0x2cf09)
    #3 0x7fa83f493343  (/usr/lib/libc++abi.so.1+0x10343)
    #4 0x7fa83f4ae932  (/usr/lib/libc++abi.so.1+0x2b932)
    #5 0x7fa83f4b18f5  (/usr/lib/libc++abi.so.1+0x2e8f5)
    #6 0x7fa83f4b1870 in __cxa_throw (/usr/lib/libc++abi.so.1+0x2e870)
    #7 0x7fa83f33ff11 in std::__throw_ios_failure(char const*) /build/gcc/src/gcc/libstdc++-v3/src/c++11/cxx11-ios_failure.cc:169:5
    #8 0x7fa83f3b4f02 in std::basic_ios<char, std::char_traits<char> >::clear(std::_Ios_Iostate) /build/gcc/src/gcc-build/x86_64-pc-linux-gnu/libstdc++-v3/include/bits/basic_ios.tcc:48:21
    #9 0x7fa83f3b8292 in std::istream::operator>>(int&) /build/gcc/src/gcc-build/x86_64-pc-linux-gnu/libstdc++-v3/include/bits/istream.tcc:171:14
    #10 0x55fa7ca735fe in main /home/wp/sync/ada-final/test.cpp:8:5
    #11 0x7fa83ef6cb24 in __libc_start_main (/usr/lib/libc.so.6+0x27b24)
    #12 0x55fa7c99135d in _start (/home/wp/sync/ada-final/a.out+0x5635d)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV /build/gcc/src/gcc/libstdc++-v3/libsupc++/typeinfo:122:32 in std::type_info::operator==(std::type_info const&) const
==347995==ABORTING
fish: Job 2, './a.out' terminated by signal SIGABRT (Abort)
```

**Additional context**

From the stack trace, it seems that `libc++abi.so` calls libstdc++'s `__do_dyncast`, which should probably not happen. Probing via gdb also shows the former feeding obviously bogus arguments to the latter.

The issue seems to be caused by `afl-clang-lto++` linking libc++ unconditionally, even when libstdc++ is used ([src/afl-cc.c#L435](https://github.com/AFLplusplus/AFLplusplus/blob/stable/src/afl-cc.c#L435)). Indeed, using the command below to build the program without AFL nor LTO but linking in libc++, the same issue persists.

```
clang++ test.cpp -lc++
```

OTOH, either commenting out line 435 or using libc++ in place of libstdc++ (`-stdlib=libc++`) fixes this issue.

Line 435 mentions that it is needed by fuzzbench, though I don't know why that is the case.

BTW, searching for similar issues brought up <https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=188088>, though their solution was just not to link in both libraries.

**Comment by vanhauser-thc** [comment]
Yes I can see the bug and understand. 
not sure we can get rid of -lc++, will have to test this thoroughly.
Thanks for reporting!

**Comment by vanhauser-thc** [comment]
tested all fuzzbench targets and they all now compile without -lc++ needed so I can remove it from the added flags.
fixed in #866.
thanks!
